baseplate/core.py

```diff
@@ -123,6 +124,145 @@ def from_upstream(cls, trace_id, parent_id, span_id, sampled, flags):
         return cls(trace_id, parent_id, span_id, sampled, flags)
 
 
+class AuthenticationContextFactory(object):
+    """Building factory for :py:class:`AuthenticationContext` objects, handles
+    context dependency for the secrets store.
+
+    This factory should be passed into the constructor of any upstream
+    :doc:`integrations <integration/index>` so that it is aware of the
+    application's secret store and can create the authentication context for
+    the incoming requests.
+
+    :param baseplate.secrets.SecretsStore secrets_store: the application's
+        defined secrets store, where application secret lookups should be made
+    """
+    def __init__(self, secrets_store=None):
+        self.secrets = secrets_store
+
+    def make_context(self, token):
+        """:py:class:`AuthenticationContext` builder
+
+        :param token: JWT value originating from the Authentication service
+            either directly or from an upstream service
+        :rtype: :py:class:`AuthenticationContext`
+        """
+        return AuthenticationContext(token=token, secrets=self.secrets)
+
+
+class AuthenticationContext(object):
+    """Wrapper for the contextual authentication information
+
+    :param str token: the JWT value for the currently propagated authentication
+        context
+    :param baseplate.secrets.SecretsStore secrets_store: the application's
+        defined secrets store, where application secret lookups should be made
+    """
+    def __init__(self, token=None, secrets=None):
+        self.token = token
+        self.secrets = secrets
+        self._payload = None
+        self._valid = None
+
+    def _secret(self):
+        if not self.secrets:
+            raise UndefinedSecretsException
+
+        return self.secrets.get_simple("jwt/authentication/secret")
+
+    def attach_context(self, context):
+        """Attach this authentication wrapper to the provided context
+
+        :param context: request context to attach this authentication to
+
+        """
+        context.authentication = self
+
+    @property
+    def valid(self):
+        """Validity of the current authentication context token
+
+        :type: bool
+        :raises: :py:class:`UndefinedSecretsException` if the
+            :py:class:`SecretsStore` has not been bound to the context handling
+            class (means that the authentication payload could not be decrypted
+            and validated)
+
+        """
+        if self._valid is not None:
+            return self._valid
+        elif self.token is None:
+            return None
+
+        try:
+            self._payload = jwt.decode(self.token, self._secret(),
+                                       algorithms='RS256')
+            self._valid = True
+        except jwt.ExpiredSignatureError:
```

Comment:
I was not *super* confident we wanted to eat all of them.  I was trying to get an idea of various failure scenarios and try to only explicitly catch them (versus something like the token getting corrupted or other wicked things) and wanted to only handle the things I had seen (even though `DecodeError` shouldn't happen for correctly designed applications, I saw it when trying to use very short RSA keys).